### PR TITLE
Upgrade-5: `prover` Refactor

### PR DESCRIPTION
**Pull request by roynalnaruto**
### Description

Refactor the `prover` crate for:
- Better code organisation
- Inline documentation
- Better structure and lesser duplication

### Notice

Requires re-running e2e tests and debugging potentially regression issues in the `prover` crate's behaviour:
1. `scroll-prover`: https://github.com/scroll-tech/scroll-prover
2. `scroll/prover`: https://github.com/scroll-tech/scroll/tree/develop/prover

### CODE REVIEW

**Comment by roynalnaruto** [comment]
@amoylan2 please refer the updates in this PR to also make corresponding changes in `scroll/prover`. @noel2004 will also be able to help.

**Comment by lispc** [comment]
> @amoylan2 please refer the updates in this PR to also make corresponding changes in `scroll/prover`. @noel2004 will also be able to help.

Btw this time we upgraded rust version and many low level libraries.  So for now better only testing this branch, later we will pick some needed changes to v0.13 for dual version support..

File: prover/src/config.rs [hunk 1]
```diff
@@ -94,43 +144,53 @@ impl LayerId {
...
-pub fn asset_file_path(filename: &str) -> String {
-    Path::new(&*ASSETS_DIR)
-        .join(filename)
-        .to_string_lossy()
-        .into_owned()
+    /// Whether or not the [`Snark`][snark_verifier_sdk::Snark] generated at this layer has an
+    /// accumulator.
+    ///
+    /// Every SNARK layer on top of the [`innermost layer`][LayerId::Inner] has an accumulator.
+    pub fn accumulator(&self) -> bool {
```
Comment (lispc) [inline prover/src/config.rs:165]: change to `has_accumulator`?

**Review by lispc — APPROVED** [review]
good refactor of Err and Proof.

can be merged after e2e testing in scroll-prover

**Comment by lispc** [comment]
i will help with e2e testing.

**Comment by lispc** [comment]
need to change 

```

/// Defines a proving task for bundle proof generation.
#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct BundleProvingTask {
    /// The [`BatchProofs`][BatchProof] for the contiguous list of batches to be bundled together.
    pub batch_proofs: Vec<BatchProof>,
}
``` 


to BatchProofV2?

File: prover/src/config.rs (hunk 1 above)
Comment (roynalnaruto) [reply prover/src/config.rs:165]: It is a planned change. I have not yet removed the `has_accumulator` argument from function calls.

**Pull request merged**